### DISCUSSION TITLE
How to use iCub velocity control with the simulator using IVelocityControl command

### DISCUSSION THREAD

**Discussion by hamaduddin** [robotology/community#587] [Q&A] [answered]
Hello,

I am a master student working on the iCub  vision more specifically on smooth pursuit and we are trying to control the icub simulator  using the velocity commands, through the 'IVelocityControl' class. i am unable to control velocity in icub by using velocityMove in python further unable to understand where velocity change effect can be seen in icub object or its eye velocity that is being affected.
this is my snippet of code :

```python
headOptions = yarp.Property()
        headOptions.put("device","remote_controlboard")
        headOptions.put("local","/client/head")
        headOptions.put("remote","/icubSim/head")
        #  create remote driver
        headDriver = yarp.PolyDriver(headOptions)
        # iMode = headDriver.viewIControlMode()
        headPos = headDriver.viewIPositionControl()
        iMode = headDriver.viewIControlMode()
        headVel = headDriver.viewIVelocityControl()
        headEncs = headDriver.viewIEncoders()
       

        joints=headPos.getAxes()
        ivelocity=headVel.velocityMove(joints,speed)
```

further these setPosition Mode and setVelocityMode commands give attribute errors in trying to implement them

```python
      for j in range(joints):
            # iMode.setPositionMode(j)
            # iMode.setPositionMode(j)
            # iMode.setVelocityMode(j)
```

Kindly provide me help on how to control it with an example of code in python ,i willl be very thankful to you in this regards.

**Comment by Nicogene** [comment, marked as answer]
Hi @hamaduddin,

welcome to the robotology community!
I don't know if the code you posted is complete but it seems that is missing some steps.

Both icub and icubSim are starting in position control, for controlling in velocity you should switch the control mode first.
Then **before** `ivelocity=headVel.velocityMove(joints,speed)` you should

```python
for j in range(joints):
    iMode.setControlMode(j, yarp.VOCAB_CM_VELOCITY)
```
In any case [here](https://github.com/robotology/blender-robotics-utils/blob/b0a87db8aa51c226588a7b4ed949a6913e3dd851/script/blenderRCBPanel/blenderRCBPanel.py#L75-L132) you can see a working example of controlling the robot using the python bindings.

Then what @pattacini said about the timeout could be a possible problem, there should be a timeout of 100 ms, if after 100 ms no velocity command is received the motion is stopped. But I expect that this is what happens on the real robot, in the simulation you should be fine, not 100% sure.
In any case, I would first fix the code and then check this thing about the timeout.
Another thing is, `speed` is a vector of the correct length?

**Reply by hamaduddin** [reply]
Ok Sir noted with thanks i will try this one .
